barcode/src/lib.rs

```diff
@@ -0,0 +1,1231 @@
+}
+
+impl<'a> TryFrom<&'a str> for Barcode<'a> {
+    type Error = BarcodeError;
+
+    /// Attempt to construct a `Barcode` instance from a string
+    fn try_from(s: &'a str) -> Result<Barcode<'a>, Self::Error> {
+        let mut split = s.split(SEPARATOR);
+
+        // the first time we `next` the iterator it will return something even
+        // if the string is empty
+        let prefix = split.next().unwrap();
+
+        // After calling `next` once, if the string returned is the same as the
+        // one provided then the pattern we `split` on isn't in the string
+        if s == prefix {
+            return Err(Self::Error::NoDelim);
+        }
+
+        let prefix = Prefix::try_from(prefix)?;
+
+        let part = split.next().ok_or(Self::Error::NoPartNumber)?;
+        let part = Part::try_from(part)?;
+
+        // the prefix determines permitted formats for the part number
+        match prefix {
+            // 0XV1 must be in v1 format
+            Prefix::ZeroXV1 => match part {
+                Part::V1(_) => (),
+                _ => return Err(Self::Error::PartNotV1),
+            },
+            // 0XV1, 0XV2 and PDV2 must be in v2 format
+            Prefix::PDV1 | Prefix::ZeroXV2 | Prefix::PDV2 => match part {
+                Part::V2(_) => (),
+                _ => return Err(Self::Error::PartNotV2),
+            },
+        }
+
+        let revision = split.next().ok_or(Self::Error::NoRevisionNumber)?;
+        let revision = Revision::try_from(revision)?;
+
+        // the prefix determines the permitted formats for the revision number
+        match prefix {
+            Prefix::PDV2 => match revision {
+                Revision::Null(_) => (),
+                _ => return Err(Self::Error::RevisionNotNull),
+            },
+            _ => match revision {
+                Revision::Number(_) => (),
+                _ => return Err(Self::Error::RevisionIsNull),
+            },
+        }
+
+        let serial = split.next().ok_or(Self::Error::NoSerialNumber)?;
+        let serial = Serial::try_from(serial)?;
+
+        // the prefix determines permitted formats for the serial number
+        match prefix {
+            // 0XV1 or PDV1: must be in v1 format
+            Prefix::ZeroXV1 | Prefix::PDV1 => match serial {
+                Serial::V1(_) => (),
+                _ => return Err(Self::Error::SerialNotV1),
+            },
+            // 0XV2 or PDV2: may be in v1 or v2 format
+            Prefix::ZeroXV2 | Prefix::PDV2 => match serial {
+                Serial::V1(_) | Serial::V2(_) => (),
+            },
+        }
+
+        Ok(Barcode {
+            prefix,
+            part,
+            revision,
+            serial,
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[derive(Debug, PartialEq)]
+pub struct BaseboardId {
+    pub part_number: String,
+    pub serial_number: String,
+}
+
+#[cfg(feature = "std")]
+#[derive(Debug, thiserror::Error, PartialEq)]
+pub enum BaseboardIdPkiPathError {
```

Comment:
Using `SlogInlineError` and `#[source]` is probably more valuable for this error type since the underlying errors are foreign and we want to display their contents.